File: nostr_client.py
import asyncio
import json
import sys

from pprint import pprint
import websockets

async def fetch_nostr_posts(relay_url="wss://relay.damus.io", limit=5):
    async with websockets.connect(relay_url) as ws:
        # Subscribe to text notes (kind 1)
        req = ["REQ", "sub1", {"kinds": [1], "limit": limit}]
        await ws.send(json.dumps(req))

        count = 0
        while count < limit:
            msg = await ws.recv()
            data = json.loads(msg)
            if data[0] == "EVENT":
                event = data[2]
                ts      = event.get("created_at")
                sender  = event.get("pubkey")[:8]  # short pubkey
                content = event.get("content")
                pprint(f"{event}\n")
                print("==============")
                count += 1

        # Close subscription
        await ws.send(json.dumps(["CLOSE", "sub1"]))

if __name__ == "__main__":
    if len(sys.argv) < 3:
            print("Usage: python nostr_client.py <relay_url> <limit>")
            print("Example: python nostr_client.py wss://relay.damus.io 5")
            sys.exit(1)
    
    asyncio.run(fetch_nostr_posts(sys.argv[1], int(sys.argv[2])))

File: node/public/clients/nostr_client.js
// nostr_client.js
// A modular Nostr client that fetches latest text notes (kind 1) from a relay
// Usage: node nostr_client.js <relay_url> [limit]
// Example: node nostr_client.js wss://relay.damus.io 5

// const ws = new WebSocket("wss://example.com");
// ws.onmessage = e => console.log(e.data);

/**
 * Fetch latest Nostr text notes (kind 1) events from a relay.
 * @param {string} relayUrl - WebSocket URL of the Nostr relay (e.g. wss://relay.damus.io)
 * @param {number} limit - Number of events to fetch (default: 5)
 * @returns {Promise<Array<Object>>} - Resolves to an array of event objects
 */
/**
 * Fetch latest Nostr text notes (kind 1) events from a relay.
 */
export function fetchNostrEvents(relayUrl, limit = 5) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(relayUrl);
    const events = [];
    const subId = 'sub1';

    ws.addEventListener('open', () => {
      ws.send(JSON.stringify(["REQ", subId, { kinds: [1], limit }]));
    });

    ws.addEventListener('message', e => {
      let msg;
      try { msg = JSON.parse(e.data) } catch { return }
      if (msg[0]==='EVENT') {
        events.push(msg[2]);
        if (events.length >= limit) {
          ws.send(JSON.stringify(["CLOSE", subId]));
          ws.close();
          resolve(events);
        }
      }
    });

    ws.addEventListener('error', reject);
    ws.addEventListener('close', () => {
      if (events.length < limit) resolve(events);
    });
  });
}


/**
 * Print Nostr events to console.
 * @param {Array<Object>} events - Array of Nostr event objects
 */
function printEvents(events) {
  if (!events.length) {
    console.log('No events found.');
    return;
  }
  events.forEach((event) => {
    const ts = event.created_at;
    const pubkey = event.pubkey ? event.pubkey.slice(0, 8) : 'unknown';
    const content = event.content || '';
    console.log(`${ts} | ${pubkey}: ${content}\n`);
  });
}

/**
 * CLI entry point
 */
async function main() {
  const [,, relayUrl, rawLimit] = process.argv;
  if (!relayUrl) {
    console.error('Usage: node nostr_client.js <relay_url> [limit]');
    process.exit(1);
  }
  const limit = Number(rawLimit) || 5;
  try {
    const events = await fetchNostrEvents(relayUrl, limit);
    printEvents(events);
  } catch (err) {
    console.error('Error fetching Nostr events:', err.message);
  }
}

// Run CLI if invoked directly
// if (require.main === module) {
//   main();
// }

// module.exports = { fetchNostrEvents, printEvents };
